crates/component-macro/src/lib.rs

```diff
@@ -0,0 +1,382 @@
+use proc_macro2::{Literal, TokenStream, TokenTree};
+use quote::quote;
+use syn::{parse_macro_input, parse_quote, Data, DeriveInput, Error, Result};
+
+static INCOMPATIBLE_TYPE_MESSAGE: &str =
+    "a `record` must be defined as a `struct` with at least one named field";
+
+#[derive(Debug)]
+enum Style {
+    Record,
+    Variant,
+    Enum,
+    Union,
+}
+
+fn find_style(input: &DeriveInput) -> Result<Style> {
+    static UNEXPECTED_FLAGS_MESSAGE: &str =
+        "`flags` not allowed here; use `wasmtime::component::flags!` macro to define `flags` types";
+
+    static UNEXPECTED_STYLE_MESSAGE: &str =
+        "unrecognized component type keyword (expected `record`, `variant`, `enum`, or `union`)";
+
+    static UNEXPECTED_SYNTAX_MESSAGE: &str = "expected `(<style>)` syntax";
+
+    let mut style = None;
+
+    for attribute in &input.attrs {
+        if attribute.path.leading_colon.is_none() && attribute.path.segments.len() == 1 {
```

Comment:
Personally for patterns like this I find it convenient to have "early return" to avoid nested indentation getting too unwieldy. For example:

```rust
if attribute.path.leading_colon.is_some() || attribute.path.segments.len() != 1 {
    continue;
}

// ... everything else minus one indentation level
```